### ISSUE TITLE
Email – Upgrade 4.4.10 to 4.44.444-beta.5 : failed to apply migration

### ISSUE THREAD

**Issue by SanboKyo** [barrelstrength/sprout#314]
### Description

After upgrading Craft (3.9.11 to 4.8.1) and all plugins (Sprout Email: 4.4.10 to 4.44.444-beta.5) all migrations were applied, except from Sprout Email (m240227_201516_schema_4_44_446).

> Exception: SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'emailTypeUid' cannot be null
> The SQL being executed was: INSERT INTO `craft_sprout_emails` (`id`, `subjectLine`, `defaultMessage`, `emailTypeUid`, `dateCreated`, `dateUpdated`, `uid`, `emailVariantType`, `emailVariantSettings`, `mailerInstructionsSettings`) VALUES ...

### How to reproduce

1. Upgrade Sprout Email from 4.4.10 to 4.44.444-beta.5
2. Run `php craft migrate/all`

### Sprout Version

4.44.444-beta.5

### Craft Version

4.8.1

### Database Type Version

MySQL 8.0.36

**Comment by BenParizek** [comment]
@SanboKyo At first glance, it appears this may be happening when Sprout Email tries to migrate the email element table and that we're expecting a value from the old database in the `emailTemplateId` field.

In the short run, you can probably solve this by making sure you have a _Template_ selected in the Notification Email settings for each of your emails in the original db before the migration.

Let me know if that works. I'll work on a more permanent solution in to add to the migrations and update this thread once I have that out.

**Comment by SanboKyo** [comment]
@BenParizek  Thx! Unfortunately I now get:

> SQLSTATE[42S02]: Base table or view not found: 1146 Table 'mhd35.craft_sprout_datasets' doesn't exist

**Comment by BenParizek** [comment]
I've just released v4.44.444-beta.6 which should resolve that.

**Comment by SanboKyo** [comment]
@BenParizek v4.44.444-beta.6 resolved that, but now I get an TypeError when I click on the Email menu item in CP:


> array_slice(): Argument #1 ($array) must be of type array, false given
> 
> in /barrelstrength/sprout/src/core/controllers/web/SettingsController.php
> 
> line 89: $firstSubNavItem = array_slice($subNav[0], 0, 1);

**Comment by BenParizek** [comment]
@SanboKyo I'm struggling to recreate this one so I must not be considering a scenario.

Can you confirm:

1. What is the URL when you get this error message? (i.e. `/admin/sprout/email/`)
2. Can you can navigate directly to `/admin/sprout/settings/control-panel`? And if so, what modules do you see enabled?
3. Are you using any other plugins or modules that update or change the CP sidebar navigation?

**Comment by SanboKyo** [comment]
@BenParizek 

1. Yes, the URL is `/admin/sprout/email/`.
2. Yes! Sent Email is enabled, Transactional is disabled.
3. Nope!

**Comment by BenParizek** [comment]
I've just pushed up 4.44.444-beta.7 which hopefully resolves this.

I wasn't able to get the same error as you, but was able to create something similar and think this improves on the logic happening there. Let me know if you continue to run into anything.

**Comment by SanboKyo** [comment]
Thx, everything seems to be working now! 

Short question: All my transactional emails had the same (the only) Email Type. I had to create new Email Types, but I couldn't edit the Email Type field of the transactional emails and so I had to create them again. Is that how it should be? That would be very inconvenient ...

**Comment by BenParizek** [comment]
@SanboKyo Each Email Type has to make assumptions about what fields are available in the Field Layout and how those available fields are handled in the Email Template. And once an email is sent, revisiting any older email may have content that is only reflected in the Email Type selected at the time.

I expect there are some additional features we can explore to support a better workflow with these over time, but right now those are the basic reasons for that restriction.

However, I can see that if you have a situation where you want to change lots of existing emails to a new type, doing so one by one could be annoying, and having some way to do them in bulk could be helpful with the right warnings.

What was the need that required you to change Email Types? Did the migration potentially miss an existing Custom Email Type that you had that you had to recreate?

**Comment by SanboKyo** [comment]
@BenParizek It was probably due to the order in which I proceeded after the update?

After updating to 4.44.444-beta.x there were no Email Types and the Transactional module was disabled. I created an Email Type and then I enabled the Transactional module. After that, every Transactional Email has the same (and only) Email Type. Then I created two more Email Types (I have 3 templates), but couldn't assign them to the Transactional Emails. That's why I had to create them again.
